tls/s2n_client_hello.c

```diff
@@ -433,9 +433,29 @@ int s2n_process_client_hello(struct s2n_connection *conn)
     /* Find potential certificate matches before we choose the cipher. */
     POSIX_GUARD(s2n_conn_find_name_matching_certs(conn));
 
+    /* Save the previous cipher suite */
+    uint8_t previous_cipher_suite_iana[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+    POSIX_CHECKED_MEMCPY(previous_cipher_suite_iana, conn->secure.cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN);
+
     /* Now choose the ciphers we have certs for. */
     POSIX_GUARD(s2n_set_cipher_as_tls_server(conn, client_hello->cipher_suites.data, client_hello->cipher_suites.size / 2));
 
+    /* Check if this is the second client hello in a hello retry handshake */
+    if (s2n_is_hello_retry_handshake(conn) &&
+        s2n_conn_get_current_message_type(conn) == CLIENT_HELLO &&
+        conn->handshake.message_number > 0) {
```

Comment:
This is in the client hello handler, so you don't need to verify that it's a client hello.

```suggestion
        conn->handshake.message_number > 0) {
```

Reply:
Fixed: eda117d